--- linux-2.6.34/drivers/mtd/nand/nand_ids.c	2011-07-21 11:21:09.000000000 +0800
+++ linux-2.6.34-new/drivers/mtd/nand/nand_ids.c	2011-07-28 16:16:06.000000000 +0800
@@ -90,6 +90,7 @@
 	{"NAND 128MiB 1,8V 16-bit",	0xB1, 0, 128, 0, LP_OPTIONS16},
 	{"NAND 128MiB 3,3V 16-bit",	0xC1, 0, 128, 0, LP_OPTIONS16},
 	{"NAND 128MiB 1,8V 16-bit",     0xAD, 0, 128, 0, LP_OPTIONS16},
+	{"NAND 128MiB 3,3V 8-bit",	0xD1, 0, 128, 0, LP_OPTIONS},
 
 	/* 2 Gigabit */
 	{"NAND 256MiB 1,8V 8-bit",	0xAA, 0, 256, 0, LP_OPTIONS},
