--- linux-2.6.34.orig/drivers/net/LIPP_6300ETH/src/gmac_drv.c	2012-04-04 11:06:38.933704084 +0900
+++ linux-2.6.34/drivers/net/LIPP_6300ETH/src/gmac_drv.c	2012-04-04 11:12:21.143712603 +0900
@@ -1974,7 +1974,7 @@
         napi_enable(&priv->napiInfo.napi);
 #endif
 
-    //netif_carrier_on(dev);
+    netif_carrier_on(dev);
     netif_start_queue( dev);
 
     GMAC_PRINT_DBG("lipp_6300Eth_open: up_lipp_6300Eth Done\n");
