--- source3/client/client.c.org	2014-05-21 18:59:39.198937484 +0900
+++ source3/client/client.c	2014-05-22 13:33:25.651803458 +0900
@@ -61,6 +61,7 @@
 
 static int process_tok(char *tok);
 static int cmd_help(void);
+static bool gShareOnly = false;
 
 #define CREATE_ACCESS_READ READ_CONTROL_ACCESS
 
@@ -3996,6 +3997,10 @@
 	/* FIXME: If the remote machine returns non-ascii characters
 	   in any of these fields, they can corrupt the output.  We
 	   should remove them. */
+	if (gShareOnly) {
+		d_printf("#%s\n", name);
+		return;
+	}
 	if (!grepable) {
 		d_printf("\t%-15s %-10.10s%s\n",
                		name,typestr,comment);
@@ -4066,7 +4071,7 @@
 static bool browse_host(bool sort)
 {
 	int ret;
-	if (!grepable) {
+	if (!grepable && !gShareOnly) {
 	        d_printf("\n\tSharename       Type      Comment\n");
 	        d_printf("\t---------       ----      -------\n");
 	}
@@ -4104,7 +4109,7 @@
 {
 	fstring state;
 
-	if (!cli->server_domain)
+	if (!cli->server_domain || gShareOnly)
 		return false;
 
 	if (!grepable) {
@@ -5121,6 +5126,7 @@
 		{ "port", 'p', POPT_ARG_INT, &port, 'p', "Port to connect to", "PORT" },
 		{ "grepable", 'g', POPT_ARG_NONE, NULL, 'g', "Produce grepable output" },
                 { "browse", 'B', POPT_ARG_NONE, NULL, 'B', "Browse SMB servers using DNS" },
+                { "shared only", 'Y', POPT_ARG_NONE, NULL, 'Y', "shared directory output only" },
 		POPT_COMMON_SAMBA
 		POPT_COMMON_CONNECTION
 		POPT_COMMON_CREDENTIALS
@@ -5252,9 +5258,11 @@
 		case 'e':
 			smb_encrypt=true;
 			break;
+		case 'Y':
+			gShareOnly=true;
+			break;
 		case 'B':
 			return(do_smb_browse());
-
 		}
 	}
 
