diff --git a/src/journal/journalctl.c b/src/journal/journalctl.c
index 4d46beb..4485eeb 100644
--- a/src/journal/journalctl.c
+++ b/src/journal/journalctl.c
@@ -675,7 +675,7 @@ static int setup_keys(void) {
         if (ioctl(fd, FS_IOC_GETFLAGS, &attr) < 0)
                 log_warning("FS_IOC_GETFLAGS failed: %m");
 
-        attr |= FS_SECRM_FL|FS_NODUMP_FL|FS_SYNC_FL|FS_NOCOW_FL;
+        attr |= FS_SECRM_FL|FS_NODUMP_FL|FS_SYNC_FL;
 
         if (ioctl(fd, FS_IOC_SETFLAGS, &attr) < 0)
                 log_warning("FS_IOC_SETFLAGS failed: %m");
