diff -Naur u-boot-2009.08/common/cmd_fat.c u-boot-2009.08_filesizefix/common/cmd_fat.c
--- u-boot-2009.08/common/cmd_fat.c	2009-08-31 23:27:42.000000000 +0530
+++ u-boot-2009.08_filesizefix/common/cmd_fat.c	2010-12-09 17:51:27.348421000 +0530
@@ -32,6 +32,7 @@
 #include <part.h>
 #include <fat.h>
 
+extern ulong NetBootFileXferSize; /*Added because this is used by by flash write routine as size*/
 
 int do_fat_fsload (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
 {
@@ -82,6 +83,8 @@
 	sprintf(buf, "%lX", size);
 	setenv("filesize", buf);
 
+        NetBootFileXferSize = size;
+
 	return 0;
 }
 
