diff -Naurp u-boot-2009.08.org/common/cmd_bootm.c u-boot-2009.08/common/cmd_bootm.c
--- u-boot-2009.08.org/common/cmd_bootm.c	2009-08-31 12:57:42.000000000 -0500
+++ u-boot-2009.08/common/cmd_bootm.c	2010-11-02 23:42:59.000000000 -0500
@@ -340,8 +340,10 @@ static int bootm_load_os(image_info_t os
 			printf ("   Loading %s ... ", type_name);
 
 			if (load != image_start) {
+                                dcache_enable();
 				memmove_wd ((void *)load,
 						(void *)image_start, image_len, CHUNKSZ);
+                                dcache_disable();
 			}
 		}
 		*load_end = load + image_len;
@@ -349,8 +351,10 @@ static int bootm_load_os(image_info_t os
 		break;
 	case IH_COMP_GZIP:
 		printf ("   Uncompressing %s ... ", type_name);
+                dcache_enable();
 		if (gunzip ((void *)load, unc_len,
 					(uchar *)image_start, &image_len) != 0) {
+                        dcache_disable();
 			puts ("GUNZIP: uncompress, out-of-mem or overwrite error "
 				"- must RESET board to recover\n");
 			if (boot_progress)
@@ -358,6 +362,7 @@ static int bootm_load_os(image_info_t os
 			return BOOTM_ERR_RESET;
 		}
 
+                dcache_disable();
 		*load_end = load + image_len;
 		break;
 #ifdef CONFIG_BZIP2
@@ -368,10 +373,12 @@ static int bootm_load_os(image_info_t os
 		 * use slower decompression algorithm which requires
 		 * at most 2300 KB of memory.
 		 */
+                dcache_enable();
 		int i = BZ2_bzBuffToBuffDecompress ((char*)load,
 					&unc_len, (char *)image_start, image_len,
 					CONFIG_SYS_MALLOC_LEN < (4096 * 1024), 0);
 		if (i != BZ_OK) {
+                        dcache_disable();
 			printf ("BUNZIP2: uncompress or overwrite error %d "
 				"- must RESET board to recover\n", i);
 			if (boot_progress)
@@ -379,6 +386,7 @@ static int bootm_load_os(image_info_t os
 			return BOOTM_ERR_RESET;
 		}
 
+                dcache_disable();
 		*load_end = load + unc_len;
 		break;
 #endif /* CONFIG_BZIP2 */
@@ -386,15 +394,20 @@ static int bootm_load_os(image_info_t os
 	case IH_COMP_LZMA:
 		printf ("   Uncompressing %s ... ", type_name);
 
+                dcache_enable();
+
 		int ret = lzmaBuffToBuffDecompress(
 			(unsigned char *)load, &unc_len,
 			(unsigned char *)image_start, image_len);
+
 		if (ret != SZ_OK) {
+                        dcache_disable();
 			printf ("LZMA: uncompress or overwrite error %d "
 				"- must RESET board to recover\n", ret);
 			show_boot_progress (-6);
 			return BOOTM_ERR_RESET;
 		}
+                dcache_disable();
 		*load_end = load + unc_len;
 		break;
 #endif /* CONFIG_LZMA */
