diff -Naur u-boot-2009.08/lib_generic/string.c u-boot-2009.08.fix/lib_generic/string.c
--- u-boot-2009.08/lib_generic/string.c	2009-08-31 12:57:42.000000000 -0500
+++ u-boot-2009.08.fix/lib_generic/string.c	2010-09-22 15:54:53.118757464 -0500
@@ -437,6 +437,7 @@
 #endif
 
 #ifndef __HAVE_ARCH_MEMCPY
+/* This was copied from the 2010.09-rc2 version of u-boot - jtm */
 /**
  * memcpy - Copy one area of memory to another
  * @dest: Where to copy to
@@ -446,12 +447,23 @@
  * You should not use this function to access IO space, use memcpy_toio()
  * or memcpy_fromio() instead.
  */
-void * memcpy(void * dest,const void *src,size_t count)
+void * memcpy(void *dest, const void *src, size_t count)
 {
-	char *tmp = (char *) dest, *s = (char *) src;
+	unsigned long *dl = (unsigned long *)dest, *sl = (unsigned long *)src;
+	char *d8, *s8;
 
+	/* while all data is aligned (common case), copy a word at a time */
+	if ( (((ulong)dest | (ulong)src) & (sizeof(*dl) - 1)) == 0) {
+		while (count >= sizeof(*dl)) {
+			*dl++ = *sl++;
+			count -= sizeof(*dl);
+		}
+	}
+	/* copy the reset one byte at a time */
+	d8 = (char *)dl;
+	s8 = (char *)sl;
 	while (count--)
-		*tmp++ = *s++;
+		*d8++ = *s8++;
 
 	return dest;
 }
